### ISSUE TITLE
Trying to estimate known isotropic Debeye Waller values for Nickel via CartNet

### ISSUE THREAD

**Issue by ZacharyVarley** [imatge-upc/CartNet#3]
Per the title I wanted to match an entry found here:

Peng, L-M., et al. "Debye–Waller factors and absorptive scattering factors of elemental crystals." Foundations of Crystallography 52.3 (1996): 456-470.

https://journals.iucr.org/a/issues/1996/03/00/zh0008/zh0008_82472sup1.pdf

From the table, the isotropic Debye Waller factor for Nickel is 0.3559 A^2 at 293 K. I wanted to see if I could estimate this using your results with CartNet which provides a richer 3x3 ellipsoid. I loaded the Nickel CIF with my stand alone file that I placed into my cloned CartNet repo. I also grabbed your ADP checkpoint from https://zenodo.org/records/13970823.

""

    import torch
    import numpy as np
    from pathlib import Path
    import sys
    from models.master import create_model
    from torch_geometric import seed_everything
    from torch_geometric.graphgym.config import cfg, set_cfg
    from torch_geometric.graphgym.logger import set_printing

    def setup_config_for_ni_inference():
        """Setup configuration for Ni FCC inference."""
        # Set basic configuration
        set_cfg(cfg)

        # Configure for inference
        cfg.seed = 0
        cfg.name = "Ni_inference"
        cfg.run_dir = "results/Ni_inference/0"
        cfg.dataset.task_type = "regression"
        cfg.batch = 1
        cfg.batch_accumulation = 1
        cfg.dataset.name = "ADP"
        cfg.dataset_path = "./Ni_sym.cif"
        cfg.loss = "MAE"
        cfg.model = "CartNet"
        cfg.max_neighbours = -1
        cfg.radius = 5.0
        cfg.num_layers = 4
        cfg.dim_in = 256
        cfg.dim_rbf = 64
        cfg.augment = False
        cfg.invariant = False
        cfg.use_temp = True
        cfg.standarize_temp = True
        cfg.envelope = True
        cfg.use_H = True
        cfg.use_atom_types = True
        cfg.workers = 0

        # Set temperature to 293K
        cfg.temperature = 293.0

        return cfg

    def load_ni_cif_data():
        """Load Ni.cif data manually."""
        from torch_geometric.data import Data
        import torch

        # Read the CIF file
        cif_path = "Ni.cif"
        if not Path(cif_path).exists():
            raise FileNotFoundError(f"Ni.cif not found at {cif_path}")

        print(f"Loading {cif_path}...")

        # Parse CIF manually using ASE
        from ase.io import read

        atoms = read(cif_path)

        # Get atomic numbers and positions
        atomic_numbers = torch.tensor(atoms.get_atomic_numbers(), dtype=torch.long)
        positions = torch.tensor(atoms.get_positions(), dtype=torch.float32)
        cell = torch.tensor(atoms.get_cell().array, dtype=torch.float32)

        # Create non-H mask (all Ni atoms)
        non_H_mask = atomic_numbers != 1

        # Create graph edges using radius graph
        from torch_geometric.nn import radius_graph

        edge_index = radius_graph(positions, r=cfg.radius, batch=None, loop=False)

        # Calculate edge features
        edge_vectors = positions[edge_index[1]] - positions[edge_index[0]]
        edge_distances = torch.norm(edge_vectors, dim=1)
        edge_directions = edge_vectors / edge_distances.unsqueeze(1)

        # Temperature data (293K)
        temperature = torch.tensor([293.0], dtype=torch.float32)

        # Normalize temperature (using ADP dataset statistics)
        temp_mean = 192.1785
        temp_std = 81.2135
        temperature_normalized = (temperature - temp_mean) / temp_std

        # Create data object
        data = Data(
            x=atomic_numbers,
            pos=positions,
            edge_index=edge_index,
            edge_attr=edge_distances,
            cart_dist=edge_distances,
            cart_dir=edge_directions,
            cell=cell.unsqueeze(0),
            non_H_mask=non_H_mask,
            temperature_og=temperature,
            temperature=temperature_normalized,
            batch=torch.zeros(len(atomic_numbers), dtype=torch.long),
        )

        return data

    def run_ni_inference():
        """Run inference on Ni.cif at 293K and extract ADP matrices."""

        # Setup configuration
        cfg = setup_config_for_ni_inference()

        # Set printing
        set_printing()

        # Seed everything
        seed_everything(cfg.seed)

        # Check if checkpoint exists
        checkpoint_path = "cartnet_adp.ckpt"
        if not Path(checkpoint_path).exists():
            print(f"Error: Checkpoint file not found at {checkpoint_path}")
            return False

        # Set device
        device = "cuda:0" if torch.cuda.is_available() else "cpu"
        print(f"Using device: {device}")

        try:
            # Load Ni.cif data
            data = load_ni_cif_data()
            data = data.to(device)

            print(f"Structure loaded: {len(data.x)} atoms, {len(data.edge_index[0])} edges")
            print(f"Non-hydrogen atoms: {data.non_H_mask.sum().item()}")

            # Create model
            model = create_model()
            model = model.to(device)

            # Load checkpoint
            print("Loading checkpoint...")
            ckpt = torch.load(checkpoint_path, map_location=device)
            model.load_state_dict(ckpt["model_state"])
            model.eval()

            # Run inference
            print("Running inference...")
            with torch.no_grad():
                prediction, _ = model(data)

            print(f"Prediction shape: {prediction.shape}")

            # Extract ADP matrices for each atom
            adp_matrices = prediction.cpu().numpy()

            # Get atom information
            positions = data.pos[data.non_H_mask].cpu().numpy()

            print("\n" + "=" * 80)
            print("ANISOTROPIC DISPLACEMENT PARAMETERS (ADP) RESULTS")
            print("=" * 80)

            # Print results for each atom
            for i, (pos, adp_matrix) in enumerate(zip(positions, adp_matrices)):
                element = "Ni"  # All atoms are Nickel

                print(f"\nAtom {i+1}: {element} (Z=28)")
                print(f"Position: [{pos[0]:.6f}, {pos[1]:.6f}, {pos[2]:.6f}] Å")
                print(f"3x3 ADP Matrix (Å²):")
                print(
                    f"  [{adp_matrix[0,0]:8.6f}  {adp_matrix[0,1]:8.6f}  {adp_matrix[0,2]:8.6f}]"
                )
                print(
                    f"  [{adp_matrix[1,0]:8.6f}  {adp_matrix[1,1]:8.6f}  {adp_matrix[1,2]:8.6f}]"
                )
                print(
                    f"  [{adp_matrix[2,0]:8.6f}  {adp_matrix[2,1]:8.6f}  {adp_matrix[2,2]:8.6f}]"
                )

                # Calculate isotropic value (trace/3)
                trace = np.trace(adp_matrix)
                isotropic = trace / 3.0

                print(f"Trace: {trace:.6f} Å²")
                print(f"Isotropic (trace/3): {isotropic:.6f} Å²")

                # Calculate eigenvalues for additional analysis
                eigenvalues = np.linalg.eigvals(adp_matrix)
                eigenvalues = np.sort(eigenvalues)  # Sort for consistent ordering

                print(
                    f"Eigenvalues: [{eigenvalues[0]:.6f}, {eigenvalues[1]:.6f}, {eigenvalues[2]:.6f}] Å²"
                )

                # Calculate anisotropy measures
                u_max = eigenvalues[2]
                u_min = eigenvalues[0]
                anisotropy_ratio = u_max / u_min if u_min > 0 else float("inf")

                print(f"Anisotropy ratio (U_max/U_min): {anisotropy_ratio:.3f}")

                print("-" * 60)

            # Summary statistics
            print(f"\nSUMMARY:")
            print(f"Total atoms analyzed: {len(adp_matrices)}")
            all_isotropic = [np.trace(matrix) / 3.0 for matrix in adp_matrices]
            print(
                f"Average isotropic value: {np.mean(all_isotropic):.6f} ± {np.std(all_isotropic):.6f} Å²"
            )
            print(
                f"Isotropic range: {np.min(all_isotropic):.6f} - {np.max(all_isotropic):.6f} Å²"
            )

            return True

        except Exception as e:
            print(f"Error during inference: {e}")
            import traceback

            traceback.print_exc()
            return False

    if __name__ == "__main__":
        success = run_ni_inference()
        if success:
            print("\n✓ Ni.cif inference completed successfully!")
        else:
            print("\n✗ Ni.cif inference failed!")
            sys.exit(1)

""

The above inference script yields:

```
Atom 1: Ni (Z=28)
Position: [0.000000, 0.000000, 0.000000] Å
3x3 ADP Matrix (Å²):
  [0.209628  -0.009202  -0.024115]
  [-0.009202  0.169170  -0.009178]
  [-0.024115  -0.009178  0.180220]
Trace: 0.559018 Å²
Isotropic (trace/3): 0.186339 Å²
Eigenvalues: [0.155272, 0.180335, 0.223410] Å²
Anisotropy ratio (U_max/U_min): 1.439
------------------------------------------------------------

Atom 2: Ni (Z=28)
Position: [0.000000, 1.737573, 1.737573] Å
3x3 ADP Matrix (Å²):
  [0.328817  0.064101  0.071475]
  [0.064101  0.222056  -0.038790]
  [0.071475  -0.038790  0.294745]
Trace: 0.845618 Å²
Isotropic (trace/3): 0.281873 Å²
Eigenvalues: [0.157835, 0.297447, 0.390336] Å²
Anisotropy ratio (U_max/U_min): 2.473
------------------------------------------------------------

Atom 3: Ni (Z=28)
Position: [1.737573, 0.000000, 1.737573] Å
3x3 ADP Matrix (Å²):
  [0.224431  0.028233  -0.018450]
  [0.028233  0.192264  0.034584]
  [-0.018450  0.034584  0.225396]
Trace: 0.642091 Å²
Isotropic (trace/3): 0.214030 Å²
Eigenvalues: [0.154274, 0.240601, 0.247217] Å²
Anisotropy ratio (U_max/U_min): 1.602
------------------------------------------------------------

Atom 4: Ni (Z=28)
Position: [1.737573, 1.737573, 0.000000] Å
3x3 ADP Matrix (Å²):
  [0.223305  -0.010311  0.045821]
  [-0.010311  0.210371  0.048006]
  [0.045821  0.048006  0.301104]
Trace: 0.734780 Å²
Isotropic (trace/3): 0.244927 Å²
Eigenvalues: [0.171592, 0.227876, 0.335313] Å²
Anisotropy ratio (U_max/U_min): 1.954
------------------------------------------------------------

SUMMARY:
Total atoms analyzed: 4
Average isotropic value: 0.231792 ± 0.035574 Å²
Isotropic range: 0.186339 - 0.281873 Å²

```

If I am interpreting it correctly, the 4 equivalent atoms have different shaped ellipsoids and the estimate is around 0.232 instead of 0.356. I greatly appreciate any insights or suggestions you may have to offer. I tried the hugging face website, but uploading Ni.cif yielded an error:  "Error: 'No such item: _atom_site_aniso_u_11'".

My goal is to get Nickel working so that I can try some more difficult materials.

**Comment by alexsoleg** [comment]
Dear @ZacharyVarley,

I think the issue is related how you are creating the graph, since the function that you are using do not take into account the periodic boundary conditions. 

```python
# Create graph edges using radius graph
    from torch_geometric.nn import radius_graph

    edge_index = radius_graph(positions, r=cfg.radius, batch=None, loop=False)
```

Can you please try again with this implementation?

[Radius Graph PBC](https://huggingface.co/spaces/alexsoleg/cartnet-demo/blob/main/utils.py)

Also the edges cannot be computed as directly subtraction of the positions since it has to take into account the cell offset. Check this [implementation](https://huggingface.co/spaces/alexsoleg/cartnet-demo/blob/main/process.py) also.


Regarding the "Error: 'No such item: _atom_site_aniso_u_11'" issue, can you provide me the CIF file that you are using? I will try to debug it.

EDIT: Can you try the huggin face now? I just realized that one of the latests push were wrong, I have fix it.

Best regards,

Àlex

**Comment by ZacharyVarley** [comment]
If I put either the symmetrized on non-symmetrized Nickel file from here (have to be logged in to export the cif):

https://next-gen.materialsproject.org/materials/mp-23/

Which is a CIF file that look like this: 

```
# generated using pymatgen
data_Ni
_diffrn_ambient_temperature 293.0
_symmetry_space_group_name_H-M   'P 1'
_cell_length_a   3.47514587
_cell_length_b   3.47514587
_cell_length_c   3.47514587
_cell_angle_alpha   90.00000000
_cell_angle_beta   90.00000000
_cell_angle_gamma   90.00000000
_symmetry_Int_Tables_number   1
_chemical_formula_structural   Ni
_chemical_formula_sum   Ni4
_cell_volume   41.96808136
_cell_formula_units_Z   4
loop_
 _symmetry_equiv_pos_site_id
 _symmetry_equiv_pos_as_xyz
  1  'x, y, z'
loop_
 _atom_type_symbol
 _atom_type_oxidation_number
  Ni0+  0.0
loop_
 _atom_site_type_symbol
 _atom_site_label
 _atom_site_symmetry_multiplicity
 _atom_site_fract_x
 _atom_site_fract_y
 _atom_site_fract_z
 _atom_site_occupancy
  Ni0+  Ni0  1  0.00000000  0.00000000  0.00000000  1
  Ni0+  Ni1  1  0.00000000  0.50000000  0.50000000  1
  Ni0+  Ni2  1  0.50000000  0.00000000  0.50000000  1
  Ni0+  Ni3  1  0.50000000  0.50000000  0.00000000  1
```

Where I manually added the temperature line "_diffrn_ambient_temperature 293.0" myself in a text editor. When I upload it to the demo on huggingface I get the following out:

```
CIF file successfully read.

Graph successfully created.

Succesfully predicted 0.00 % of ADPs

CartNet produced unexpected ADPs for the following atoms (will be ignored in the output file): Ni1, Ni2, Ni3, Ni4

ADPs successfully predicted.
```

I get the same result if I use the longer CIF that uses space group 225 and lists the related fields " _symmetry_equiv_pos_site_id" and "_symmetry_equiv_pos_as_xyz". Thank you again for your time and for your patience with my sluggish response.

**Comment by alexsoleg** [comment]
I think this problem arise with the fact that, for the thermal ellipsoids, we trained CartNet solely on molecular crystals from the CSD, and your input is a material not a molecular crystal. 

If you are interested, and you know any option to get thermal ellipsoid data from materials, I can run a finetuning or something to make it work also for materials.

NOTE: The message "Successfully predicted 0.00 % of ADPs" is a sanity check that I added for the cases if the model outputs some atoms that are no physically correct to remove them but leave the others. In this case, seems that all of them are bad, possibly because the model never seen a structure like that. If you want to get rid of this message and see what the model is giving you even though are wrong, download locally the hugging face app and comment the lines where we do this filtering.

**Comment by ZacharyVarley** [comment]
Àlex, 

Thanks for your time. That makes sense that it is out of distribution. I will keep an eye out for models that have been trained on materials / geological data. I am not sure if there are thermal ellipsoid data for crystalline materials. If it comes via DFT then I think there may be enough training examples out there but I am not certain. I will let you know what I find. I am closing this issue. Thanks again for your time and for answering my questions.

-Zach
